File: array/array6.go
/*
Given two arrays, write a function to compute their intersection.

Example 1:

Input: nums1 = [1,2,2,1], nums2 = [2,2]
Output: [2,2]
Example 2:

Input: nums1 = [4,9,5], nums2 = [9,4,9,8,4]
Output: [4,9]
Note:

Each element in the result should appear as many times as it shows in both arrays.
The result can be in any order.
Follow up:

What if the given array is already sorted? How would you optimize your algorithm?
What if nums1's size is small compared to nums2's size? Which algorithm is better?
What if elements of nums2 are stored on disk, and the memory is limited such that you cannot load all elements into the memory at once?
*/

package array

// Intersect Intersection of Two Arrays II
func Intersect(nums1 []int, nums2 []int) []int {
	sumArray := make([]int, 0)
	compareArray := make([]int, 0)

	save := map[int]int{}
	list := make([]int, 0)

	// temp = nums1
	// nums2 > nums1 => temp = nums2
	sumArray = nums1
	compareArray = nums2
	if len(nums1) < len(nums2) {
		sumArray = nums2
		compareArray = nums1
	}

	for _, value := range compareArray {
		save[value]++
	}

	for _, value := range sumArray {
		if save[value] > 0 {
			list = append(list, value)
			save[value]--
		}
	}
	return list
}

File: old/Array/array6.js
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number[]}
 */
var intersect = function (nums1, nums2) {
  const intersection = [];

  const small = nums1.length < nums2.length ? nums1 : nums2;
  const big = small === nums1 ? nums2 : nums1;

  small.sort((a, b) => a - b);
  big.sort((a, b) => a - b);

  // 각각의 배열에 포인터 생성
  let i = 0;
  let j = 0;
  while (i < small.length) {
    // intersection 찾으면 push 하고 포인터 모두 이동
    if (small[i] === big[j]) {
      intersection.push(small[i]);
      i++;
      j++;
      continue;
    }
    // intersection 아닐 경우 크기 비교해 작은 쪽의 포인터만 이동
    small[i] > big[j] ? j++ : i++;
  }

  return intersection;
};
